### PR TITLE
feature(state-processor): send only validators set diffs to consensus

### PR DESCRIPTION
**Pull request by abi87**
Whenever we turn the epoch, we must inform consensus of the changes in the validator set composition.
It's best to send only diffs of the validator set wrt previous epoch, rather than sending the whole new validator set.
This is what this PR does.

## Summary by CodeRabbit

- **Documentation**
	- Updated the documentation for validator handling to clarify conditions for returning validators to the consensus engine.

- **Bug Fixes**
	- Improved clarity of error messages related to proposer mismatches.

- **New Features**
	- Introduced a mechanism to track previous epoch validators, enhancing the efficiency of validator updates.
	- Added a new method for processing validator set updates, optimizing the consensus mechanism.

- **Refactor**
	- Modified methods to improve the handling of validator updates and streamline processing logic.

- **Tests**
	- Simplified assertions in test cases related to validator transitions, enhancing clarity and focus.

### CODE REVIEW

File: mod/state-transition/pkg/core/state_processor.go [hunk 1]
```diff
@@ -94,6 +94,11 @@ type StateProcessor[
 	// processingGenesis allows initializing correctly
 	// eth1 deposit index upon genesis
 	processingGenesis bool
+
+	// prevEpochValidators tracks the set of validators active during
+	// previous epoch. This is useful at the turn of the epoch to send to
+	// consensus only the diffs rather than the full updated validator set.
+	prevEpochValidators []*transition.ValidatorUpdate
```
Comment (abi87) [inline mod/state-transition/pkg/core/state_processor.go:101]: note that we only keep active validators here, i.e. those whose effective balance is larger than ejection one.
I do not expect this list to be beefy, although we should probably track its size and decide whether this is the best way to calculate diffs.

A less memory consuming approach is to track chances wherever we do them (e.g.  when updating effective balances, or creating validators or removing it). I prefer current approach is it is more easily verifiable for correctness for me, but we should measure how much of a performance improvement these two alternatives bring

File: mod/state-transition/pkg/core/state_processor_genesis.go [hunk 2]
```diff
@@ -166,5 +166,5 @@ func (sp *StateProcessor[
 		return nil, err
 	}
 
-	return sp.processSyncCommitteeUpdates(st)
```
Comment (abi87) [inline mod/state-transition/pkg/core/state_processor_genesis.go:169]: `processSyncCommitteeUpdates` does not really make sense to me.
It seems to me we're trying to squeeze a pure BeaconKit operation (updating CometBFT val set) into an Eth 2.0 one

File: mod/state-transition/pkg/core/state_processor_staking_test.go [hunk 3]
```diff
@@ -214,15 +213,7 @@ func TestTransitionUpdateValidator(t *testing.T) {
...
-	require.Len(t, newEpochVals, len(genDeposits)) // just topped up one validator
-
-	// Assuming genesis order is preserved here which is not necessary
-	// TODO: remove this assumption
-
-	// all genesis validators other than the last are unchanged
-	for i := range len(genDeposits) - 1 {
-		require.Equal(t, genVals[i], newEpochVals[i], fmt.Sprintf("idx: %d", i))
-	}
+	require.Len(t, newEpochVals, 1) // just topped up one validator
```
Comment (abi87) [inline mod/state-transition/pkg/core/state_processor_staking_test.go:216]: only send diffs

File: mod/state-transition/pkg/core/state_processor_staking_test.go [hunk 4]
```diff
@@ -400,15 +391,7 @@ func TestTransitionCreateValidator(t *testing.T) {
...
-	require.Len(t, newEpochVals, len(genDeposits)+1)
-
-	// Assuming genesis order is preserved here which is not necessary
-	// TODO: remove this assumption
-
-	// all genesis validators are unchanged
-	for i := range len(genDeposits) {
-		require.Equal(t, genVals[i], newEpochVals[i], fmt.Sprintf("idx: %d", i))
-	}
+	require.Len(t, newEpochVals, 1) // just added 1 validator
```
Comment (abi87) [inline mod/state-transition/pkg/core/state_processor_staking_test.go:394]: only send diff.

File: mod/state-transition/pkg/core/helpers_test.go [hunk 5]
```diff
@@ -219,3 +220,18 @@ func buildNextBlock(
...
+	)
+
+	dummyExecutionPayload = &types.ExecutionPayload{
+		Timestamp:     0,
+		ExtraData:     []byte("testing"),
+		Transactions:  [][]byte{},
+		Withdrawals:   []*engineprimitives.Withdrawal{},
+		BaseFeePerGas: math.NewU256(0),
+	}
+)
```
Comment (abi87) [inline mod/state-transition/pkg/core/helpers_test.go:237]: minor UT consolidation

File: mod/state-transition/pkg/core/state_processor_validators.go [hunk 6]
```diff
@@ -58,4 +60,45 @@ func (sp *StateProcessor[
...
+
+	// prevValsSet now contains all evicted validators (and only those)
+	for pkBytes := range prevValsSet {
+		//#nosec:G703 // bytes comes from a pk
+		pk, _ := bytes.ToBytes48([]byte(pkBytes))
+		res = append(res, &transition.ValidatorUpdate{
+			Pubkey:           pk,
+			EffectiveBalance: 0, // signal val eviction to consensus
+		})
+	}
```
Comment (abi87) [inline mod/state-transition/pkg/core/state_processor_validators.go:99]: we do not evict validators yet, so this should not really happen. Still added for completeness

**Review by calbera — APPROVED** [review]
logically makes sense, agree with trying to optimize this in future

**Review by nidhi-singh02 — APPROVED** [review]

File: mod/state-transition/pkg/core/state_processor_validators.go (hunk 6 above)
Comment (abi87) [reply mod/state-transition/pkg/core/state_processor_validators.go:99]: in fact we may evict them as soon as we introduce a cap on the validators set

**Review by fridrik01 — APPROVED** [review]
LGTM

File: mod/state-transition/pkg/core/state_processor_validators.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+	res, err := sp.validatorSetsDiffs(prevEpochVals, activeVals)
+	if err != nil {
+		return nil, err
+	}
+
+	// clear up sets we won't lookup to anymore
+	sp.valSetByEpoch[currEpoch] = activeVals
+	if prevEpoch >= 1 {
+		delete(sp.valSetByEpoch, prevEpoch-1)
+	}
```
[the remark this answers is not in the document]
Comment (abi87) [reply mod/state-transition/pkg/core/state_processor_validators.go:74]: Epochs are not skipped

File: mod/state-transition/pkg/core/state_processor_validators.go [hunk 8]
```diff
@@ -58,4 +60,45 @@ func (sp *StateProcessor[
...
+		}
+
+		// consume pre-existing validators
+		delete(prevValsSet, key)
+	}
+
+	// prevValsSet now contains all evicted validators (and only those)
+	for pkBytes := range prevValsSet {
+		//#nosec:G703 // bytes comes from a pk
+		pk, _ := bytes.ToBytes48([]byte(pkBytes))
```
[the remark this answers is not in the document]
Comment (abi87) [reply mod/state-transition/pkg/core/state_processor_validators.go:127]: no need, it's fine

**Pull request merged**